File: backend/rpc/handler/user/management.py
from rpc.gen.user.management.services import TUserManagementService
from rpc.gen.user.user.types.ttypes import TUserRole
from rpc.gen.user.user.structs.ttypes import TUser
from rpc.gen.user.management.errors.ttypes import TUserManagementError, TUserManagementErrorCode
from ..crud import TCrudServiceHandler

from models import models

class TUserManagementServiceHandler(TUserManagementService.Iface, TCrudServiceHandler):
    def __init__(self):
        TUserManagementService.Iface.__init__(self)
        TCrudServiceHandler.__init__(
            self,
            "user", 
            TUser,
            actions={
                "fetch": TUserRole.ADMIN_AKUN,
                "delete": TUserRole.ADMIN_AKUN,
            },
            setters={
                "name": TUserRole.SUPER_ADMIN,
                "role": TUserRole.ADMIN_AKUN,
                "email": TUserRole.SUPER_ADMIN,
                "password": TUserRole.SUPER_ADMIN,
                "enabled": TUserRole.ADMIN_AKUN,
                "verified": TUserRole.SUPER_ADMIN,
            },
            require_role=True
        )
        self.email_model = models['email']

    # def set_email(self, auth_token, id, email):
    #     raise NotImplementedError()

    def create(self, auth_token, form):
        actor = self.auth_model.require_role(auth_token, TUserRole.ADMIN_AKUN)
        db_user = self.model.create(form, actor=actor)
        self.email_model.send_welcome(db_user)
        self.model.commit()
        user = db_user.fill(TUser())
        return user

File: frontend/src/rpc/client/user/management.js
import TUserManagementService from '@/rpc/gen/TUserManagementService';
import { TCrudClient } from '@/rpc/client/crud';
import { TUserRole } from '@/rpc/gen/user.user.types_types';


class TUserManagementClient extends TCrudClient{
	constructor(stores=null){
		super(
			stores, 
			TUserManagementService, 
			'/api/user/management',
			{
				fetch: TUserRole.ADMIN_AKUN,
				create: TUserRole.ADMIN_AKUN,
				delete: TUserRole.SUPER_ADMIN,
			},
			{}
		);
	}

	async set_role(id, value){
		await this.set_field("role", id, value, TUserRole.ADMIN_AKUN);
		if (id == this.stores.auth.user.id){
			//await this.stores.auth.setUserRole(new_role);
			await this.stores.client.user.auth.logout();
		}
	}

	async set_email(id, value){
		await this.set_field("email", id, value, TUserRole.ADMIN_AKUN);
		if (id == this.stores.auth.user.id){
			await this.stores.client.user.auth.logout();
		}
	}
	async set_enabled(id, value){
		await this.set_field("enabled", id, value, TUserRole.ADMIN_AKUN);
		if (id == this.stores.auth.user.id && !value){
			await this.stores.client.user.auth.logout();
		}
	}
	async set_password(id, value){
		await this.set_field("password", id, value, TUserRole.SUPER_ADMIN);
		if (id == this.stores.auth.user.id){
			// await this.stores.client.user.auth.logout();
			await this.stores.client.user.auth.login(this.stores.auth.user.username, new_password);
		}
	}
	async set_verified(id, value){
		await this.set_field("verified", id, value, TUserRole.SUPER_ADMIN);
		if (id == this.stores.auth.user.id && !value){
			await this.stores.client.user.auth.logout();
		}
	}
	async delete(id){
		await super.delete(id);
		// this.requireRoleAction("delete");
		// await this.client.delete(this.stores.auth.authToken, id);
		if (id == this.stores.auth.user.id){
			await this.stores.client.user.auth.logout();
		}
	}
}

const userManagementClient = new TUserManagementClient();

export { TUserManagementClient, userManagementClient }
export default userManagementClient